monkey/monkey_island/cc/services/reporting/report.py

```diff
@@ -240,83 +261,126 @@ def get_island_cross_segment_issues():
+                            "hostname": machine.hostname,
+                            "target": str(target_ip),
+                            "services": None,
+                            "is_self": True,
+                        }
+                    )
 
         return cross_segment_issues
 
-    @staticmethod
-    def get_cross_segment_issues_per_subnet_pair(scans, source_subnet, target_subnet):
+    @classmethod
+    def get_cross_segment_issues_per_subnet_pair(
+        cls,
+        scans: Sequence[Union[PingScanEvent, TCPScanEvent]],
+        source_subnet: str,
+        target_subnet: str,
+    ) -> List[Dict[str, Any]]:
         """
         Gets list of cross segment issues from source_subnet to target_subnet.
 
-        :param scans: List of all scan telemetry entries. Must have monkey_guid,ip_addr and
-         services. This should be a PyMongo cursor object.
+        :param scans: List of all successful scan events.
         :param source_subnet:   The subnet which shouldn't be able to access target_subnet.
         :param target_subnet:   The subnet which shouldn't be accessible from source_subnet.
         :return:
         """
         if source_subnet == target_subnet:
             return []
+
+        if cls._agent_repository is None:
+            raise RuntimeError()
+        if cls._machine_repository is None:
+            raise RuntimeError()
+
         source_subnet_range = NetworkRange.get_range_obj(source_subnet)
         target_subnet_range = NetworkRange.get_range_obj(target_subnet)
 
         cross_segment_issues = []
 
-        scans.rewind()  # If we iterated over scans already we need to rewind.
+        agents_dict = {a.id: a for a in cls._agent_repository.get_agents()}
+        machines_dict = {m.id: m for m in cls._machine_repository.get_machines()}
+        machine_events: Dict[Machine, Dict[IPv4Address, Dict[Type, ScanEvent]]] = defaultdict(
+            lambda: defaultdict(lambda: defaultdict())
+        )
+
+        # Store events for which the target IP is in the target subnet, indexed
+        # by the scanning machine, target IP, and event type
         for scan in scans:
-            target_ip = scan["data"]["machine"]["ip_addr"]
+            target_ip = scan.target
             if target_subnet_range.is_in_range(str(target_ip)):
-                monkey = NodeService.get_monkey_by_guid(scan["monkey_guid"])
-                cross_segment_ip = get_ip_in_src_and_not_in_dst(
-                    monkey["ip_addresses"], source_subnet_range, target_subnet_range
-                )
+                agent = agents_dict[scan.source]
+                machine = machines_dict[agent.machine_id]
+                machine_events[machine][target_ip][type(scan)] = scan
+
+        # Report issues when the machine has an IP in the source subnet range
+        for machine, scan_dict in machine_events.items():
+            machine_ips = [iface.ip for iface in machine.network_interfaces]
```

Comment:
Seems like this issue is just a subset of ` get_cross_segment_issues_of_single_machine`? Because it checks for the same thing (machines have IP's in both subnets) but this also checks for any communications? I'm not sure how valuable that is.

Reply:
You're right, it's fairly similar. I suppose it makes a bit of a stronger case in that it was able to successfully scan a machine